cache/gradle.go

```diff
@@ -0,0 +1,129 @@
+	   Gradle 6.1.1
+	   ------------------------------------------------------------
+
+	   Build time:   2020-01-24 22:30:24 UTC
+	   Revision:     a8c3750babb99d1894378073499d6716a1a1fa5d
+
+	   Kotlin:       1.3.61
+	   Groovy:       2.5.8
+	   Ant:          Apache Ant(TM) version 1.10.7 compiled on September 1 2019
+	   JVM:          1.8.0_241 (Oracle Corporation 25.241-b07)
+	   OS:           Mac OS X 10.15.5 x86_64
+	*/
+
+	pattern := `Gradle (.*)`
+	exp := regexp.MustCompile(pattern)
+	matches := exp.FindStringSubmatch(out)
+	if len(matches) < 2 {
+		return "", fmt.Errorf("failed to find Gradle version in output:\n%s\nusing a pattern:%s", out, pattern)
+	}
+	return matches[1], nil
+}
+
+func projectGradleVersion(projectPth string) (string, error) {
+	gradlewPth := filepath.Join(projectPth, "gradlew")
+	exist, err := pathutil.IsPathExists(gradlewPth)
+	if err != nil {
+		return "", fmt.Errorf("failed to check if %s exists: %s", gradlewPth, err)
+	}
+	if !exist {
+		return "", fmt.Errorf("no gradlew found at: %s", gradlewPth)
+	}
+
+	versionCmd := command.New("./gradlew", "-version")
+	versionCmd.SetDir(filepath.Dir(gradlewPth))
+	out, err := versionCmd.RunAndReturnTrimmedCombinedOutput()
+	if err != nil {
+		if errorutil.IsExitStatusError(err) {
+			return "", fmt.Errorf("%s failed: %s", versionCmd.PrintableCommandArgs(), out)
+		}
+		return "", fmt.Errorf("%s failed: %s", versionCmd.PrintableCommandArgs(), err)
+	}
+
+	return parseGradleVersion(out)
+}
+
+func oldGradleExcludePaths(homeDir, currentGradleVersion string) ([]string, error) {
+	var excludes []string
+
+	gradlewDir := filepath.Join(homeDir, ".gradle")
+	exist, err := pathutil.IsPathExists(gradlewDir)
+	if err != nil {
+		return nil, fmt.Errorf("failed to check if %s exist: %s", gradlewDir, err)
+	}
+	if !exist {
+		return nil, nil
+	}
+
+	{
+		// exclude old wrappers, like ~/.gradle/wrapper/dists/gradle-5.1.1-all
+		wrapperDistrDir := filepath.Join(gradlewDir, "wrapper", "dists")
+		entries, err := ioutil.ReadDir(wrapperDistrDir)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read entries of %s: %s", wrapperDistrDir, err)
+		}
+		for _, e := range entries {
+			if !strings.HasPrefix(e.Name(), "gradle-"+currentGradleVersion) {
+				excludes = append(excludes, "!"+filepath.Join(wrapperDistrDir, e.Name()))
+			}
+		}
+	}
+
+	{
+		// exclude old caches, like ~/.gradle/caches/5.1.1
+		cachesDir := filepath.Join(gradlewDir, "caches")
+		entries, err := ioutil.ReadDir(cachesDir)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read entries of %s: %s", cachesDir, err)
+		}
+		for _, e := range entries {
+			v, err := version.NewVersion(e.Name())
+			if err != nil || v == nil {
+				continue
+			}
+
+			if e.Name() != currentGradleVersion {
+				excludes = append(excludes, "!"+filepath.Join(cachesDir, e.Name()))
+			}
+		}
+	}
+
+	{
+		// exclude old daemon, like ~/.gradle/daemon/5.1.1
+		daemonDir := filepath.Join(gradlewDir, "daemon")
```

Comment:
Interested if there are any not gradle version directories in the daemon directory?

Reply:
I am not aware of any, this is how it looks on my machine:

```
.
├── 2.13
├── 4.1
├── 5.1.1
├── 5.4.1
├── 5.6.2
├── 5.6.4
├── 6.1.1
└── 6.3
```

Reply:
Was thinking about if an include list can be used in that case instead of an exclude. But not seeing benefit for that (the only case would be if parsing gradle version fails), so can go with the current solution.